File: libs/statemanager/src/main/kotlin/mono/state/ClipboardManager.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state

import mono.graphics.geo.Point
import mono.lifecycle.LifecycleOwner
import mono.shape.clipboard.ShapeClipboardManager
import mono.shape.connector.ShapeConnector
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group
import mono.state.command.CommandEnvironment

/**
 * A manager class to handle clipboard related data.
 */
internal class ClipboardManager(
    lifecycleOwner: LifecycleOwner,
    private val environment: CommandEnvironment,
    private val shapeClipboardManager: ShapeClipboardManager
) {
    private var selectedShapes: Collection<AbstractShape> = emptyList()

    init {
        environment.selectedShapesLiveData.observe(lifecycleOwner) {
            selectedShapes = it
        }
        shapeClipboardManager.clipboardShapeLiveData.observe(
            lifecycleOwner,
            listener = ::pasteShapes
        )
    }

    fun copySelectedShapes(isRemoveRequired: Boolean) {
        shapeClipboardManager.setClipboard(createClipboardObject())
        if (isRemoveRequired) {
            for (shape in selectedShapes) {
                environment.removeShape(shape)
            }
            environment.clearSelectedShapes()
        }
    }

    private fun pasteShapes(clipboardObject: ShapeClipboardManager.ClipboardObject) {
        if (clipboardObject.shapes.isEmpty()) {
            return
        }
        environment.clearSelectedShapes()
        val bound = environment.getWindowBound()
        val left = bound.left + bound.width / 5
        val top = bound.top + bound.height / 5
        insertShapes(left, top, clipboardObject)
    }

    fun duplicateSelectedShapes() {
        if (selectedShapes.isEmpty()) {
            return
        }
        val currentSelectedShapes = selectedShapes
        val clipboardObject = createClipboardObject()
        val minLeft = currentSelectedShapes.minOf { it.bound.left }
        val minTop = currentSelectedShapes.minOf { it.bound.top }

        environment.clearSelectedShapes()
        insertShapes(minLeft + 1, minTop + 1, clipboardObject)
    }

    private fun insertShapes(
        left: Int,
        top: Int,
        clipboardObject: ShapeClipboardManager.ClipboardObject
    ) {
        val currentParentId = environment.workingParentGroup.id

        val srcIdToShapeMap =
            clipboardObject.shapes.associate { it.id to Group.toShape(currentParentId, it) }
        val minLeft = srcIdToShapeMap.values.minOf { it.bound.left }
        val minTop = srcIdToShapeMap.values.minOf { it.bound.top }

        val offset = Point(minLeft - left, minTop - top)
        for (shape in srcIdToShapeMap.values) {
            val shapeBound = shape.bound
            val newShapeBound = shapeBound.copy(position = shapeBound.position.minus(offset))
            shape.setBound(newShapeBound)

            environment.addShape(shape)
            environment.addSelectedShape(shape)
        }

        for (connector in clipboardObject.connectors) {
            val line = srcIdToShapeMap[connector.lineId] as? mono.shape.shape.Line ?: continue
            val target = srcIdToShapeMap[connector.targetId] ?: continue
            environment.shapeManager.shapeConnector.addConnector(line, connector.anchor, target)
        }
    }

    private fun createClipboardObject(): ShapeClipboardManager.ClipboardObject {
        val serializableShapes = selectedShapes.map { it.toSerializableShape(false) }
        val serializableConnectors = selectedShapes.flatMap { target ->
            environment.shapeManager.shapeConnector.getConnectors(target).map {
                ShapeConnector.toSerializableConnector(it, target)
            }
        }
        return ShapeClipboardManager.ClipboardObject(serializableShapes, serializableConnectors)
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/clipboard-manager.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { LifecycleOwner } from "$libs/flow";
import { Point } from "$libs/graphics-geo/point";
import { Rect } from "$libs/graphics-geo/rect";
import { ShapeConnector } from "$mono/shape/connector/shape-connector";
import { ClipboardObject, type ShapeClipboardManager } from "$mono/shape/shape-clipboard-manager";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import { Group } from "$mono/shape/shape/group";
import { Line } from "$mono/shape/shape/line";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";

/**
 * A manager class to handle clipboard related data.
 */
export class ClipboardManager {
    private selectedShapes: AbstractShape[] = [];

    constructor(
        private environment: CommandEnvironment,
        private shapeClipboardManager: ShapeClipboardManager,
    ) {
    }

    observe(lifecycleOwner: LifecycleOwner) {
        this.environment.selectedShapesFlow.observe(lifecycleOwner, (shapes) => {
            this.selectedShapes = Array.from(shapes);
        });
        this.shapeClipboardManager.clipboardShapeFlow.observe(
            lifecycleOwner,
            this.pasteShapes.bind(this),
        );
    }

    copySelectedShapes(isRemoveRequired: boolean) {
        this.shapeClipboardManager.setClipboard(this.createClipboardObject());
        if (isRemoveRequired) {
            for (const shape of this.selectedShapes) {
                this.environment.removeShape(shape);
            }
            this.environment.clearSelectedShapes();
        }
    }

    private pasteShapes(clipboardObject: ClipboardObject) {
        if (clipboardObject.shapes.length === 0) {
            return;
        }
        this.environment.clearSelectedShapes();
        const bound = this.environment.getWindowBound();
        const left = bound.left + bound.width / 5;
        const top = bound.top + bound.height / 5;
        this.insertShapes(left, top, clipboardObject);
    }

    duplicateSelectedShapes() {
        if (this.selectedShapes.length === 0) {
            return;
        }
        const currentSelectedShapes = this.selectedShapes;
        const clipboardObject = this.createClipboardObject();
        const minLeft = Math.min(...currentSelectedShapes.map(shape => shape.bound.left));
        const minTop = Math.min(...currentSelectedShapes.map(shape => shape.bound.top));

        this.environment.clearSelectedShapes();
        this.insertShapes(minLeft + 1, minTop + 1, clipboardObject);
    }

    private insertShapes(
        left: number,
        top: number,
        clipboardObject: ClipboardObject,
    ) {
        const currentParentId = this.environment.workingParentGroup.id;

        const srcIdToShapeMap = new Map(
            clipboardObject.shapes.map(shape => [shape.id, Group.toShape(currentParentId, shape)]),
        );
        const minLeft = Math.min(...Array.from(srcIdToShapeMap.values()).map(shape => shape.bound.left));
        const minTop = Math.min(...Array.from(srcIdToShapeMap.values()).map(shape => shape.bound.top));

        const offset = Point.from({ left: minLeft - left, top: minTop - top });
        for (const shape of srcIdToShapeMap.values()) {
            const shapeBound = shape.bound;
            const newShapeBound = new Rect(shapeBound.position.minus(offset), shapeBound.size);
            shape.setBound(newShapeBound);

            this.environment.addShape(shape);
            this.environment.addSelectedShape(shape);
        }

        for (const connector of clipboardObject.connectors) {
            const line = srcIdToShapeMap.get(connector.lineId);
            if (!line || !(line instanceof Line)) {
                continue;
            }
            const target = srcIdToShapeMap.get(connector.targetId);
            if (line && target) {
                this.environment.shapeManager.shapeConnector.addConnector(line, connector.anchor, target);
            }
        }
    }

    private createClipboardObject(): ClipboardObject {
        const serializableShapes =
            this.selectedShapes.map(shape => shape.toSerializableShape(false));
        const serializableConnectors = this.selectedShapes.flatMap(target =>
            this.environment.shapeManager.shapeConnector.getConnectors(target).map(connector =>
                ShapeConnector.toSerializableConnector(connector, target),
            ),
        );
        return ClipboardObject.create(serializableShapes, serializableConnectors);
    }
}
